nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/lossfunctions/SameDiffLoss.java

```diff
@@ -0,0 +1,156 @@
+        INDArray output = activationFn.getActivation(preOutput.dup(), true);
+
+        Map<String, INDArray> m = new HashMap<>();
+        m.put("labels", this.sd.getVariable("labels").eval());
+        m.put("layerInput", this.sd.getVariable("layerInput").eval());
+
+
+        // How scoreArr var appear in graph? is it in someway connected to output?  ScoreArr is the same as lossVal
+        // So its depends on output, have spaghetti understanding
+        INDArray scoreArr = sd.output(m,"scoreArr").get("scoreArr");
+
+
+        if (mask != null) {
+            LossUtil.applyMask(scoreArr, mask);
+        }
+        return scoreArr;
+
+
+    }
+
+
+    /**
+     * Compute the gradient of the loss function with respect to the inputs: dL/dOutput
+     *
+     * @param labels       Label/expected output
+     * @param preOutput    Output of the model (neural network), before the activation function is applied
+     * @param activationFn Activation function that should be applied to preOutput
+     * @param mask         Mask array; may be null
+     * @return Gradient dL/dPreOut
+     */
+    public INDArray computeGradient(INDArray labels, INDArray preOutput, IActivation activationFn, INDArray mask) {
+
+
+
+        Map<String, INDArray> m = new HashMap<>();
+        m.put("labels", this.sd.getVariable("labels").eval());
+        m.put("layerInput", this.sd.getVariable("layerInput").eval());
+
+        Map<String,INDArray> grads = sd.calculateGradients(m, "layerInput");
+
+        if (mask != null) {
+            LossUtil.applyMask(grads.get("layerInput"), mask);
+        }
+        return grads.get("layerInput");
+    }
+
+    /**
+     * Compute both the score (loss function value) and gradient. This is equivalent to calling {@link #computeScore(INDArray, INDArray, IActivation, INDArray, boolean)}
+     * and {@link #computeGradient(INDArray, INDArray, IActivation, INDArray)} individually
+     *
+     * @param labels       Label/expected output
+     * @param preOutput    Output of the model (neural network)
+     * @param activationFn Activation function that should be applied to preOutput
+     * @param mask         Mask array; may be null
+     * @param average      Whether the score should be averaged (divided by number of rows in labels/output) or not
+     * @return The score (loss function value) and gradient
+     */
+
+
+    public  Pair<Double, INDArray> computeGradientAndScore(INDArray labels, INDArray preOutput, IActivation activationFn,
+                                                   INDArray mask, boolean average) {
+
+        Pair<Double, INDArray> GradientAndScore = new Pair<>();
+        GradientAndScore.setFirst(this.computeScore(labels, preOutput, activationFn, mask, average));
+        GradientAndScore.setSecond(this.computeGradient(labels, preOutput, activationFn, mask));
+
+        return GradientAndScore;
+    }
+
+    public String name() {
+        return toString();
```

Comment:
Let's change this to `return getClass().getSimpleName()`
Currently it's going to call the default Object.toString() method